include/boost/beast/websocket/error.hpp

```diff
@@ -20,22 +21,230 @@ namespace websocket {
+
+        Error codes with this value will compare equal to @ref condition::handshake_failed
+    */
+    bad_sec_version,
+
+    /** The WebSocket handshake Sec-WebSocket-Accept field is missing
+
+        Error codes with this value will compare equal to @ref condition::handshake_failed
+    */
+    no_sec_accept,
+
+    /** The WebSocket handshake Sec-WebSocket-Accept field is invalid
+
+        Error codes with this value will compare equal to @ref condition::handshake_failed
+    */
+    bad_sec_accept,
+
+    /** The WebSocket handshake was declined by the remote peer
+
+        Error codes with this value will compare equal to @ref condition::handshake_failed
+    */
+    upgrade_declined,
+
+    //
+    // Protocol errors
+    //
+    // These will compare equal to condition::protocol_violation
+    //
+
+    /** The WebSocket frame contained an illegal opcode
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_opcode,
+
+    /** The WebSocket data frame was unexpected
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_data_frame,
+
+    /** The WebSocket continuation frame was unexpected
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_continuation,
+
+    /** The WebSocket frame contained illegal reserved bits
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_reserved_bits,
+
+    /** The WebSocket control frame was fragmented
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_control_fragment,
+
+    /** The WebSocket control frame size was invalid
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_control_size,
+
+    /** The WebSocket frame was unmasked
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_unmasked_frame,
+
+    /** The WebSocket frame was masked
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_masked_frame,
+
+    /** The WebSocket frame size was not canonical
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_size,
+
+    /** The WebSocket frame payload was not valid utf8
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_frame_payload,
+
+    /** The WebSocket close frame reason code was invalid
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_close_code,
+
+    /** The WebSocket close frame payload size was invalid
+
+        Error codes with this value will compare equal to @ref condition::protocol_violation
+    */
+    bad_close_size,
```

Comment:
Looking at this and the next, they appear to match some close_code values? My initial reaction is that `websocket::close_reason::code` should be returning a close_code enum, with its own category specifically for close_code. Then the error_condition could still compare equal for `condition::protocol_violation` on those codes, or any other close_codes that mean that. Users could also then tell if they have a valid close_reason from looking at the category of the error_code they received.

Reply:
Close codes only represent a subset of the possible errors that can take place. And sometimes the value is not set correctly by the remote host. So the error code provides more concise information. I don't think it is a good idea to try to hook the close code system into a category. Especially because there can be custom close codes (for example in the 3,000 to 3,999 range) which are application specific and may have different meanings.

Reply:
The "subset of possible errors" is why you'd have a category. The category decides how to deal with the interactions of that subset with the real world. Application specific close codes could easily work by deriving from a `ws_close_code_category`, and handling the 3XXX codes, and then upcalling for anything else.

The idea of the category is that if you have different meanings for error numbers (i.e. close_code 7 != `boost::beast::websocket::error` 7, then having different categories lets the category author define the interoperation.

That said, close codes (e.g. none) might fall into that somewhat grey area where "non-zero success and/or zero failure" is not clearly supported or disallowed, and there's an ongoing need to clarify this.

I also see that these two error codes are not set based on the peer-provided close code, but determined locally, so my idea of replacing these is not right. The general idea that close_code fits in the error_code model however, I do think is valid, modulo the above paragraph.

It's also possible this demonstrates my under-experience with Beast itself.

Reply:
I think it is more subjective. For example there is a group who thinks that the error code system should be used to reflect HTTP status codes (e.g. 200 OK). I am not in that group.

Reply:
Indeed, I'm one of those folks. That was my point about "non-zero success", as that turns out to be one of the technical sticking points to letting the implementors decide that; specifically, the semantic of `error_code::operator bool` is not clear, and usage guidance in blog-posts conflicts over it (as well as whether HTTP status codes make sense in `error_code`).

Lack of message/string/data support in `error_code` is also pretty clear, and various approaches show up in projects like Boost.outcome (gives you a pointer into a ring-buffer) or our local `error_code` usage (locally subclassed `std::error_code` with a string added, and we just accept that slicing to `std::error_code` will lose that detail), to carry those around. It seems that as far as the standardisation effort was concerned, that's what the `what` in `system_error` was for, and if you weren't going to put your `error_code` into `system_error` and throw it, carrying a specific string was just left off-the-table. FIlesystem TS similarly gives richer results when throwing `path_error` than when simply setting an `error_code`, so this is not a once-off choice.

Anyway, this isn't really a discussion for here. `error_code` practice is still evolving, I know of at least two people who've indicated plans to write papers addressing issues like this, and I also have a (smaller) one planned, but far down my backlog.